File: funcscript/block/SelectorExpression.cs
using FuncScript.Core;
using FuncScript.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;

namespace FuncScript.Block
{
    internal class SelectorExpression: ExpressionBlock
    {
        class SelectorProvider: IFsDataProvider
        {
            public IFsDataProvider Provider;
            public SelectorExpression Parent;
            public IFsDataProvider ParentProvider => Provider;

            public object SourceVal
            {
                set
                {
                    _sourceVal = value as KeyValueCollection;
                }
            }
            KeyValueCollection _sourceVal;
            public object Get(string name)
            {
                if (_sourceVal != null)
                {
                    if (_sourceVal.IsDefined(name))
                        return _sourceVal.Get(name);
                }
                return Provider.Get(name);
            }
            public bool IsDefined(string key)
            {
                if (_sourceVal != null)
                {
                    if (_sourceVal.IsDefined(key))
                        return true;
                }

                return Provider.IsDefined(key);
            }

        }
        public ExpressionBlock Source;
        public KvcExpression Selector;
        public override object Evaluate(IFsDataProvider provider)
        {
            var sourceVal = Source.Evaluate(provider);
            if (sourceVal is FsList)
            {
                var lst = (FsList)sourceVal;
                var ret = new object[lst.Length];
                int i = 0;
                
                foreach (var l in lst)
                {
                    var sel=new SelectorProvider
                    {
                        Parent = this,
                        Provider = provider,
                        SourceVal = l
                    };
                    ret[i] = Selector.Evaluate(sel);
                    i++;
                }
                return new ArrayFsList(ret);

            }
            else
            {
                return Selector.Evaluate(new SelectorProvider
                {
                    Parent = this,
                    Provider = provider,
                    SourceVal=sourceVal
                });
            }
        }

        public override IList<ExpressionBlock> GetChilds()
        {
            return new ExpressionBlock[] { Source, Selector };
        }
        public override string ToString()
        {
            return "selector";
        }
        public override string AsExpString(IFsDataProvider provider)
        {
            return $"{Source.AsExpString(provider)} {Selector.AsExpString(provider)}";
        }
    }
}

File: js_port/fs_js/src/block/SelectorExpression.js
const { ExpressionBlock } = require('./ExpressionBlock');
const { FsDataProvider } = require('../core/dataProvider');
const { ensureTyped, typeOf, valueOf, makeValue, typedNull } = require('../core/value');
const { FSDataType } = require('../core/fstypes');
const { ArrayFsList } = require('../model/FsList');

class SelectorProvider extends FsDataProvider {
  constructor(parent) {
    super(parent);
    this.sourceValue = typedNull();
    this.sourceCollection = null;
  }

  setSource(value) {
    const typed = ensureTyped(value);
    this.sourceValue = typed;
    if (typeOf(typed) === FSDataType.KeyValueCollection) {
      this.sourceCollection = valueOf(typed);
    } else {
      this.sourceCollection = null;
    }
  }

  get(name) {
    const lower = name.toLowerCase();
    if (this.sourceCollection && this.sourceCollection.isDefined(lower)) {
      return this.sourceCollection.get(lower);
    }
    return super.get(name);
  }

  isDefined(name) {
    const lower = name.toLowerCase();
    if (this.sourceCollection && this.sourceCollection.isDefined(lower)) {
      return true;
    }
    return super.isDefined(name);
  }
}

class SelectorExpression extends ExpressionBlock {
  constructor() {
    super();
    this.Source = null;
    this.Selector = null;
  }

  evaluate(provider) {
    const sourceTyped = ensureTyped(this.Source.evaluate(provider));
    const sourceType = typeOf(sourceTyped);
    const selectorProvider = new SelectorProvider(provider);

    if (sourceType === FSDataType.List) {
      const list = valueOf(sourceTyped);
      const results = [];
      for (let i = 0; i < list.length; i += 1) {
        const item = list.get(i) ?? typedNull();
        selectorProvider.setSource(item);
        const result = ensureTyped(this.Selector.evaluate(selectorProvider));
        results.push(result);
      }
      return makeValue(FSDataType.List, new ArrayFsList(results));
    }

    selectorProvider.setSource(sourceTyped);
    return ensureTyped(this.Selector.evaluate(selectorProvider));
  }

  getChilds() {
    return [this.Source, this.Selector];
  }

  asExpressionString(provider) {
    return `${this.Source.asExpressionString(provider)} ${this.Selector.asExpressionString(provider)}`;
  }
}

module.exports = {
  SelectorExpression
};
